### PR TITLE
Unify mocking and overriding in E2E tests

### PR DESCRIPTION
**Pull request by anthony-murphy**
There were a couple different patterns in our e2e test to mock and override functionality, primarily for fault injection. This change unifies all the usages i could find on a common mechanism that is concise and easy to use.

### CODE REVIEW

File: packages/test/test-end-to-end-tests/src/mocking.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+export type OverrideFactory<T, P extends keyof T> = (T: T) => T[P];
+
+type NestedOverrides<T> = {
+	[P in keyof T]?: T[P] extends (...args: any) => any
+		? NestedOverrides<UnPromise<ReturnType<T[P]>>> | OverrideFactory<T, P>
+		: OverrideFactory<T, P>;
+};
+
+export function wrapObjectAndOverride<T extends Record<string, any>>(
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/mocking.ts:18]: Did you look into Sinon?  https://sinonjs.org/releases/v17/  It's very possible it offers similar functionality out of the box.

File: packages/test/test-end-to-end-tests/src/mocking.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { isPromiseLike } from "@fluidframework/core-utils";
+
+export type UnPromise<T> = T extends Promise<infer U> ? U : T;
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/mocking.ts:8]: What about just using `Awaited`?  https://www.typescriptlang.org/docs/handbook/utility-types.html

File: packages/test/test-end-to-end-tests/src/mocking.ts (hunk 1 above)
Comment (markfields) [reply packages/test/test-end-to-end-tests/src/mocking.ts:18]: I'll admit I have also found it difficult to understand/use at times, this API looks potentially more approachable.  Consider using Sinon to implement it though.

Comment (anthony-murphy) [reply packages/test/test-end-to-end-tests/src/mocking.ts:18]: briefly, and didn't see anything that looked similar. we already had this is our code base, just reusing it.

File: packages/test/test-end-to-end-tests/src/mocking.ts (hunk 2 above)
Comment (anthony-murphy) [reply packages/test/test-end-to-end-tests/src/mocking.ts:8]: nice. i think this code was written before it that existed, as this is just moved from another file to be reused

File: packages/test/test-end-to-end-tests/src/test/blobs.spec.ts [hunk 3]
```diff
@@ -703,28 +688,24 @@ describeCompat("blobs", "NoCompat", (getTestObjectProvider, apis) => {
...
-			async apply(target, thisArg, args) {
-				return uploadBlobWithDelay(target, thisArg, args);
+		const uploadBlobPromise = new Deferred<void>();
+		const container1 = await provider.makeTestContainer({
+			...testContainerConfig,
+			loaderProps: {
+				documentServiceFactory: wrapObjectAndOverride(provider.documentServiceFactory, {
+					createDocumentService: {
+						connectToStorage: {
+							createBlob: (dss) => async (blob) => {
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/blobs.spec.ts:698]: I like how explicit this is about what is being mocked and how

Comment (markfields) [reply packages/test/test-end-to-end-tests/src/test/blobs.spec.ts:698]: @pragya91 if you are still looking to mock storage network calls, you can probably use `wrapObjectAndOverride` that Tony's standardizing on in this PR.

File: packages/test/test-end-to-end-tests/src/mocking.ts (hunk 1 above)
Comment (anthony-murphy) [reply packages/test/test-end-to-end-tests/src/mocking.ts:18]: i didn't add this in this PR, just copy pasted from packages/test/test-end-to-end-tests/src/test/batchBreakRegression.spec.ts

**Review by markfields — APPROVED** [review]
LGTM! Left a few comments to think about

File: packages/test/test-utils/api-report/test-utils.api.md [hunk 4]
```diff
@@ -393,13 +393,13 @@ export interface TimeoutWithValue<T = void> {
 // @internal
 export function waitForContainerConnection(container: IContainer, failOnContainerClose?: boolean, timeoutOptions?: TimeoutWithError): Promise<void>;
 
-// @internal
+// @internal @deprecated (undocumented)
```
[the remark this answers is not in the document]
Comment (anthony-murphy) [reply packages/test/test-utils/api-report/test-utils.api.md:396]: i don't really trust internal yet. once we start trimming, we can def do that

File: packages/test/test-end-to-end-tests/src/mocking.ts (hunk 2 above)
Comment (markfields) [reply packages/test/test-end-to-end-tests/src/mocking.ts:8]: I thought it was a more refined / flexible version of what you have here. e.g. deals with multiple / mismatched `Promise`.  Does it not work out to use it here?

File: packages/test/test-utils/api-report/test-utils.api.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (markfields) [reply packages/test/test-utils/api-report/test-utils.api.md:396]: Ohhhh I for sure thought we were already trimming. I'll follow up offline about that

File: packages/test/test-end-to-end-tests/src/mocking.ts (hunk 2 above)
Comment (anthony-murphy) [reply packages/test/test-end-to-end-tests/src/mocking.ts:8]: i'm getting type error when i try to use it, and looking a the definition i think the nesting nature it the culprit, i only want to strip off the top level promise if it exists

**Pull request merged**